Cargo.toml

```diff
@@ -28,6 +28,7 @@ rand = "0.8.5"
 rand_xorshift = "0.3.0"
 log = {  version = "0.4.22", features = ["std"] }
 once_cell = "1.19.0"
+tracing = "0.1.40"
```

Comment:
do we need to enable the "log" feature so that log records are generated?

Reply:
are you talking about the general log that we see or specific time period based logs cause then we can work with span and events (part of tracing crate).

Reply:
from https://docs.rs/tracing/latest/tracing/#emitting-log-records:

```
This crate provides two feature flags, “log” and “log-always”, which will cause [spans](https://docs.rs/tracing/latest/tracing/span/index.html) and [events](https://docs.rs/tracing/latest/tracing/event/struct.Event.html) to emit log records. When the “log” feature is enabled, if no tracing Subscriber is active, invoking an event macro or creating a span with fields will emit a log record. This is intended primarily for use in libraries which wish to emit diagnostics that can be consumed by applications using tracing or log, without paying the additional overhead of emitting both forms of diagnostics when tracing is in use.
```

It sounds like the `log` feature needs to be enabled to actually emit log records (as emitted by the basic rust logging facade (https://docs.rs/log/latest/log/)) when using `tracing`?

Reply:
yes there is a flag feature as you mentioned but also there is [this](https://docs.rs/tracing/latest/tracing/#crate-feature-flags:~:text=tracing%2Dsubscriber%20provides%20Subscriber%20implementations%20and%20utilities%20for%20working%20with%20Subscribers.%20This%20includes%20a%20FmtSubscriber%20FmtSubscriber%20for%20logging%20formatted%20trace%20data%20to%20stdout%2C%20with%20similar%20filtering%20and%20formatting%20to%20the%20env_logger%20crate) which we would want , also enabling feature of log we will have logs in text base format with field from level TRACE not from the actual level it is happening from you can see [this](https://docs.rs/tracing/latest/tracing/#emitting-log-records:~:text=.%20Since%20these%20additional%20span%20lifecycle%20logs%20have%20the%20potential%20to%20be%20very%20verbose%2C%20and%20don%E2%80%99t%20include%20additional%20fields%2C%20they%20will%20always%20be%20emitted%20at%20the%20Trace%20level%2C%20rather%20than%20inheriting%20the%20level%20of%20the%20span%20that%20generated%20them.)

Reply:
> yes there is a flag feature as you mentioned but also there is [this](https://docs.rs/tracing/latest/tracing/#crate-feature-flags:~:text=tracing%2Dsubscriber%20provides%20Subscriber%20implementations%20and%20utilities%20for%20working%20with%20Subscribers.%20This%20includes%20a%20FmtSubscriber%20FmtSubscriber%20for%20logging%20formatted%20trace%20data%20to%20stdout%2C%20with%20similar%20filtering%20and%20formatting%20to%20the%20env_logger%20crate) which we would want

Not sure what the suggestion is here. I agree you'd need to set up a subscriber to actually consume and record the logs somewhere useful. But I think that's up to the user of slatedb to import and set up. Since SlateDB is a library, it probably shouldn't set up any subscriber itself.

> also enabling feature of log we will have logs in text base format with field from level TRACE not from the actual level it is happening from you can see [this](https://docs.rs/tracing/latest/tracing/#emitting-log-records:~:text=.%20Since%20these%20additional%20span%20lifecycle%20logs%20have%20the%20potential%20to%20be%20very%20verbose%2C%20and%20don%E2%80%99t%20include%20additional%20fields%2C%20they%20will%20always%20be%20emitted%20at%20the%20Trace%20level%2C%20rather%20than%20inheriting%20the%20level%20of%20the%20span%20that%20generated%20them.)

I think all that's saying is that when the log feature is enabled, `tracing` will create log records at TRACE whenever a span is entered/exited/closed. Spans are different from events, and denote when certain contexts are entered/exited. For example, you could create a span that's entered whenever a function F is called and exited whenever the function returns. Within that span you could emit multiple events (which are the result of the calls to info!, error!, etc). It's likely that spans are generally much more frequently entered/exited than logs are emitted. E.G you may call F 1000 times but only every log once when some error happens. So that's saying the span enter/exit message will always be at TRACE level so you don't log too much.

Reply:
yes thanks for the explanation I was not suggesting but trying to understand what we would want to include and what not `Since SlateDB is a library, it probably shouldn't set up any subscriber itself.` this is the crux, we would just need minimal log feature then which happens occasionally ;)

Reply:
Getting up to speed on `tracing`. My first time using it. I agree with @rodesai. The way I read the [Emitting `log` Records](https://docs.rs/tracing/latest/tracing/#emitting-log-records) section, it sounds like we want either `log` or `log-always`:

> This is intended primarily for use in libraries which wish to emit diagnostics that can be consumed by applications using tracing or log, without paying the additional overhead of emitting both forms of diagnostics when tracing is in use.

I think this is what we want. If no consumer is set up, just default to `log`. I don't think we want `log-always`, since that would cause both `log` and `tracing` events to be emitted:

> Unlike the “log” feature, libraries generally should not enable the “log-always” feature, as doing so will prevent applications from being able to opt out of the log records.

tl;dr I think adding the `[log]` feature in `Cargo.toml` is the way to go.

Reply:
+1 to enabling `log` feature. The `log` feature in the tracing crate allows tracing to interoperate with the log crate, which is the traditional logging system in Rust.